File: cv-triloka-backend/app/Models/Invoice.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class Invoice extends Model
{
    use HasFactory;

    protected $fillable = [
        'invoice_number',
        'klien_id',
        'quotation_id',
        'created_by',
        'invoice_date',
        'due_date',
        'subtotal',
        'tax',
        'discount',
        'total',
        'status',
        'notes',
        'paid_at',
        'va_number',
        'va_bank',
        'va_expires_at',
        'invoice_type',
        'parent_invoice_id',
        'is_survey_fee_applied',
    ];

    protected $casts = [
        'invoice_date' => 'date',
        'due_date' => 'date',
        'paid_at' => 'datetime',
        'va_expires_at' => 'datetime',
        'subtotal' => 'decimal:2',
        'tax' => 'decimal:2',
        'discount' => 'decimal:2',
        'total' => 'decimal:2',
    ];

    /**
     * Client who owns this invoice
     */
    public function klien()
    {
        return $this->belongsTo(User::class, 'klien_id');
    }

    /**
     * Admin who created this invoice
     */
    public function creator()
    {
        return $this->belongsTo(User::class, 'created_by');
    }

    /**
     * Invoice items
     */
    public function items()
    {
        return $this->hasMany(InvoiceItem::class);
    }

    /**
     * Payments for this invoice
     */
    public function payments()
    {
        return $this->hasMany(Payment::class);
    }

    /**
     * Check if invoice is overdue
     */
    public function isOverdue(): bool
    {
        return $this->status !== 'paid' && $this->due_date->isPast();
    }

    /**
     * Check if invoice is paid
     */
    public function isPaid(): bool
    {
        return $this->status === 'paid';
    }

    /**
     * Calculate remaining balance
     */
    public function remainingBalance(): float
    {
        $totalPaid = $this->payments()->sum('amount');
        return $this->total - $totalPaid;
    }

    /**
     * Auto-generate invoice number and VA number
     */
    protected static function boot()
    {
        parent::boot();

        static::creating(function ($invoice) {
            // Generate invoice number
            if (!$invoice->invoice_number) {
                $invoice->invoice_number = self::generateInvoiceNumber();
            }
            
            // Generate Virtual Account number
            if (!$invoice->va_number) {
                $invoice->va_number = self::generateVANumber();
                $invoice->va_bank = 'BCA';
                $invoice->va_expires_at = now()->addDays(7);
            }
        });
    }

    /**
     * Generate unique invoice number
     */
    public static function generateInvoiceNumber(): string
    {
        $date = now()->format('Ymd');
        $count = self::whereDate('created_at', today())->count() + 1;
        return 'INV-' . $date . '-' . str_pad($count, 3, '0', STR_PAD_LEFT);
    }

    /**
     * Generate Virtual Account number (Mock)
     * Format: 8808 + invoice_id + random digits
     * Example: 8808 0001 234567
     */
    public static function generateVANumber(): string
    {
        // Get next ID (approximate)
        $nextId = self::max('id') + 1;
        
        // Generate random 6 digits
        $random = rand(100000, 999999);
        
        // Format: 8808 + 4-digit ID + 6-digit random
        // Example: 8808000123456 7 (displayed as 8808 0001 234567)
        return '8808' . str_pad($nextId, 4, '0', STR_PAD_LEFT) . $random;
    }

    /**
     * Check if this is a survey invoice
     */
    public function isSurveyInvoice(): bool
    {
        return $this->invoice_type === 'survey';
    }

    /**
     * Get the parent invoice (for project invoices that have survey fee)
     */
    public function parentInvoice()
    {
        return $this->belongsTo(Invoice::class, 'parent_invoice_id');
    }

    /**
     * Get the child invoice (for survey invoices)
     */
    public function childInvoice()
    {
        return $this->hasOne(Invoice::class, 'parent_invoice_id');
    }

    /**
     * Get standard survey fee amount
     */
    public static function getSurveyFeeAmount(): float
    {
        return 500000; // Rp 500k
    }

    /**
     * Apply survey fee discount to this invoice
     */
    public function applySurveyFeeDiscount(Invoice $surveyInvoice): void
    {
        if ($this->isSurveyInvoice()) {
            return; // Cannot apply to survey invoice itself
        }

        if ($surveyInvoice->status !== 'paid') {
            return; // Survey must be paid first
        }

        $this->parent_invoice_id = $surveyInvoice->id;
        $this->is_survey_fee_applied = true;
        $this->discount = ($this->discount ?? 0) + self::getSurveyFeeAmount();
        $this->total = $this->subtotal + $this->tax - $this->discount;
        $this->save();
    }
}

File: lib/models/invoice.dart
class Invoice {
  final int id;
  final String invoiceNumber;
  final int klienId;
  final String? klienName;
  final int createdBy;
  final String? creatorName;
  final String invoiceDate;
  final String dueDate;
  final double subtotal;
  final double tax;
  final double discount;
  final double total;
  final String status;
  final String? notes;
  final String? paidAt;
  final List<InvoiceItem> items;
  final double totalPaid;
  final double remainingBalance;
  final String createdAt;
  final String updatedAt;
  final String? vaNumber;
  final String? vaBank;
  final String? vaExpiresAt;
  final String invoiceType;
  final int? parentInvoiceId;
  final bool isSurveyFeeApplied;

  Invoice({
    required this.id,
    required this.invoiceNumber,
    required this.klienId,
    this.klienName,
    required this.createdBy,
    this.creatorName,
    required this.invoiceDate,
    required this.dueDate,
    required this.subtotal,
    required this.tax,
    required this.discount,
    required this.total,
    required this.status,
    this.notes,
    this.paidAt,
    this.items = const [],
    this.totalPaid = 0,
    this.remainingBalance = 0,
    required this.createdAt,
    required this.updatedAt,
    this.vaNumber,
    this.vaBank,
    this.vaExpiresAt,
    this.invoiceType = 'project',
    this.parentInvoiceId,
    this.isSurveyFeeApplied = false,
  });

  factory Invoice.fromJson(Map<String, dynamic> json) {
    // Helper to safely parse numbers (handles both int, double, and string)
    double _parseDouble(dynamic value) {
      if (value == null) return 0.0;
      if (value is double) return value;
      if (value is int) return value.toDouble();
      if (value is String) return double.tryParse(value) ?? 0.0;
      return 0.0;
    }

    // Parse items if available
    List<InvoiceItem> itemsList = [];
    if (json['items'] != null) {
      itemsList = (json['items'] as List)
          .map((item) => InvoiceItem.fromJson(item))
          .toList();
    }

    // Calculate totals
    double totalPaid = 0;
    if (json['payments'] != null) {
      for (var payment in json['payments']) {
        totalPaid += _parseDouble(payment['amount']);
      }
    }

    double total = _parseDouble(json['total']);
    double remainingBalance = total - totalPaid;

    return Invoice(
      id: json['id'],
      invoiceNumber: json['invoice_number'] ?? '',
      klienId: json['klien_id'],
      klienName: json['klien']?['name'],
      createdBy: json['created_by'],
      creatorName: json['creator']?['name'],
      invoiceDate: json['invoice_date'] ?? '',
      dueDate: json['due_date'] ?? '',
      subtotal: _parseDouble(json['subtotal']),
      tax: _parseDouble(json['tax']),
      discount: _parseDouble(json['discount']),
      total: total,
      status: json['status'] ?? 'unpaid',
      notes: json['notes'],
      paidAt: json['paid_at'],
      items: itemsList,
      totalPaid: totalPaid,
      remainingBalance: remainingBalance,
      createdAt: json['created_at'] ?? '',
      updatedAt: json['updated_at'] ?? '',
      vaNumber: json['va_number'],
      vaBank: json['va_bank'],
      vaExpiresAt: json['va_expires_at'],
      invoiceType: json['invoice_type'] ?? 'project',
      parentInvoiceId: json['parent_invoice_id'],
      isSurveyFeeApplied:
          json['is_survey_fee_applied'] == true ||
          json['is_survey_fee_applied'] == 1,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'invoice_number': invoiceNumber,
      'klien_id': klienId,
      'created_by': createdBy,
      'invoice_date': invoiceDate,
      'due_date': dueDate,
      'subtotal': subtotal,
      'tax': tax,
      'discount': discount,
      'total': total,
      'status': status,
      'notes': notes,
      'paid_at': paidAt,
      'created_at': createdAt,
      'updated_at': updatedAt,
    };
  }

  String getStatusLabel() {
    switch (status.toLowerCase()) {
      case 'paid':
        return 'Lunas';
      case 'unpaid':
        return 'Belum Bayar';
      case 'overdue':
        return 'Jatuh Tempo';
      case 'cancelled':
        return 'Dibatalkan';
      case 'draft':
        return 'Draft';
      default:
        return status;
    }
  }

  bool isOverdue() {
    if (status == 'paid' || status == 'cancelled') return false;
    try {
      DateTime due = DateTime.parse(dueDate);
      return due.isBefore(DateTime.now());
    } catch (e) {
      return false;
    }
  }
}

class InvoiceItem {
  final int id;
  final int invoiceId;
  final int? itemId;
  final String itemName;
  final String? description;
  final double quantity;
  final double unitPrice;
  final double subtotal;
  final String createdAt;
  final String updatedAt;

  InvoiceItem({
    required this.id,
    required this.invoiceId,
    this.itemId,
    required this.itemName,
    this.description,
    required this.quantity,
    required this.unitPrice,
    required this.subtotal,
    required this.createdAt,
    required this.updatedAt,
  });

  factory InvoiceItem.fromJson(Map<String, dynamic> json) {
    // Helper to safely parse numbers
    double _parseDouble(dynamic value) {
      if (value == null) return 0.0;
      if (value is double) return value;
      if (value is int) return value.toDouble();
      if (value is String) return double.tryParse(value) ?? 0.0;
      return 0.0;
    }

    return InvoiceItem(
      id: json['id'],
      invoiceId: json['invoice_id'],
      itemId: json['item_id'],
      itemName: json['item_name'] ?? '',
      description: json['description'],
      quantity: _parseDouble(json['quantity']),
      unitPrice: _parseDouble(json['unit_price']),
      subtotal: _parseDouble(json['subtotal']),
      createdAt: json['created_at'] ?? '',
      updatedAt: json['updated_at'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'invoice_id': invoiceId,
      'item_id': itemId,
      'item_name': itemName,
      'description': description,
      'quantity': quantity,
      'unit_price': unitPrice,
      'subtotal': subtotal,
      'created_at': createdAt,
      'updated_at': updatedAt,
    };
  }
}
